### PR TITLE
fix(AutoControlledComponent): use setState instead of an assign

### LINKED ISSUE #1682: Dropdown value not populated on initial render w/ react 16
### Steps
<Dropdown selection options={states} value='CA' placeholder='Select State' />

### Expected Result
California would be selected/active

### Actual Result
Placeholder is shown

### Version
semantic-ui-react@0.68.3
react@16.0.0-alpha.6

### Testcase
http://codepen.io/anon/pen/jmpavw

---

Upon further investigation it looks like semantic-ui-react is following some conventions not recommended by the react docs. Specifically I tracked the issue down to:

https://github.com/Semantic-Org/Semantic-UI-React/blob/v0.68.3/src/lib/AutoControlledComponent.js#L149

> Never mutate this.state directly, as calling setState() afterwards may replace the mutation you made. Treat this.state as if it were immutable.

This turns out to be a messy problem. The dropdown is dependent on `this.state` being set sync (it uses `this.state.value` in order to find the `selectedIndex`), but the state needs to be set with the async `setState` method otherwise it gets blown out on the event loop that `setState` finally fires.

There may need to be a larger discussion around how controlled components behave, but I was able to side step the problem by replacing the above line with:

```
      this.state = _extends({}, this.state, initialAutoControlledState)
      this.setState(this.state)
```

The overall idea being it can take advantage of `this.state` being set sync but it doesn't get blown out because I also add the update to the `setState` queue.

### LINKED ISSUE #1761: AutoControlledComponent throws warning in React 16
### Steps
1) Upgrade to React 16
2) Place any of the `AutoControlledComponent`s in your app

### Expected Result
It should render without warning

### Actual Result
It renders with the warning `Warning: Dropdown.componentWillMount(): Assigning directly to this.state is deprecated (except inside a component's constructor). Use setState instead.`
screen shot 2017-06-12 at 10 32 06 am

### Version
0.68.5

### Testcase
It seems to be caused by this line: `src/lib/AutoControlledComponent.js:149`. I'll open a PR shortly so we can continue to fully support React 16.

### PR DESCRIPTION
**Pull request by layershifter**
Fixes #1761.
Fixes #1682.

In #1770, I suggested that the problem is in that `setState` is an async, I was right. `AutoControlledComponent` component's children such as `Dropdown` and  `Search` had own logic inside the `componentWillMount` method that relied on the current state.

The `componentWillMount` method was renamed to `constructor` as it's a single place where `state` is assignable. However, we need to have ability to assign an initial state, so I added the `initialState` method that accepts `props` as an argument.

@apiv Thanks for you suggestions in #1770.

### CODE REVIEW

File: src/addons/Portal/Portal.js [hunk 1]
```diff
@@ -136,8 +136,6 @@ class Portal extends Component {
     type: META.TYPES.ADDON,
   }
 
-  state = {}
-
```
Comment (layershifter) [inline src/addons/Portal/Portal.js:140]: `state` is assigned inside the constructor now, cruft.

File: src/modules/Dropdown/Dropdown.js [hunk 2]
```diff
@@ -358,7 +358,6 @@ export default class Dropdown extends Component {
   static Menu = DropdownMenu
 
   componentWillMount() {
-    if (super.componentWillMount) super.componentWillMount()
```
Comment (layershifter) [inline src/modules/Dropdown/Dropdown.js:361]: Cruft.

File: src/lib/AutoControlledComponent.js [hunk 3]
```diff
@@ -68,8 +68,11 @@ export const getAutoControlledStateValue = (propName, props, state, includeDefau
 }
 
 export default class AutoControlledComponent extends Component {
-  componentWillMount() {
+  constructor(...args) {
+    super(...args)
+
     const { autoControlledProps } = this.constructor
+    const state = _.invoke(this, 'initialState', this.props) || {}
```
Comment (levithomason) [inline src/lib/AutoControlledComponent.js]: How about we go with `createReactClass`'s method name `getInitialState()` here?

Comment (layershifter) [reply src/lib/AutoControlledComponent.js]: I like, saw that you made changes :+1:

**Comment by apiv** [comment]
Ah, assumed my solution wasn't ideal and closed my PR, guess I should have left it open 😂.

Really excited to finish our migration to React 16 now that this is fixed, thanks!

**Comment by levithomason** [comment]
Released in `semantic-ui-react@0.71.0`.

**Pull request merged**